js/CobrowseIO.js

```diff
@@ -4,6 +4,112 @@ const NativeEventEmitter = require('react-native').NativeEventEmitter
 
 const emitter = new NativeEventEmitter(CobrowseIONative)
 
+class Session {
+  constructor (session) {
+    this._session = session || {}
+    this._listen()
+  }
+
+  _update (session, onEnd) {
+    if (this._session.id === undefined || (session && this._session.id === session.id)) {
+      this._session = session
+      if (typeof onEnd === 'function') onEnd()
+    }
+  }
+
+  _listen () {
+    const requests = emitter.addListener(CobrowseIONative.SESSION_REQUESTED, (session) => this._update(session))
+    const updates = emitter.addListener(CobrowseIONative.SESSION_UPDATED, (session) => this._update(session))
+    const ended = emitter.addListener(CobrowseIONative.SESSION_ENDED, (session) => {
+      this._update(session, () => {
+        requests.remove()
+        updates.remove()
+        ended.remove()
+      })
+    })
+  }
+
+  addListener (eventType, listener) {
+    return emitter.addListener(eventType, (session) => {
+      this._update(session, () => {
+        listener(this)
+      })
+    })
+  }
+
+  get id () {
+    return this._session.id
+  }
+
+  get code () {
+    return this._session.code
+  }
+
+  get state () {
+    return this._session.state
+  }
+
+  get full_device () {
+    return this._session.full_device
+  }
+
+  get remote_control () {
+    return this._session.remote_control
+  }
+
+  get agent () {
+    return this._session.agent
+  }
+
+  activate () {
+    return CobrowseIONative.activateSession().then((session) => {
+      this._session = session
+    })
```

Comment:
Is the then() actually required here? The session activation change should come back through the updated event as well

Reply:
This does guarantee that if `activate` is called, it comes back void. I'll remove the variable assignment, but leave the void function so the promise returns reliably. I don't want the class to leak the wrapped session object.